### DISCUSSION TITLE
Ny funksjonalitet i webserver

### DISCUSSION THREAD

**Discussion by heming-langrenn** [langrenn-sprint/sprint-webserver#15] [Ideas]
Hva skal vi lage av nye funksjoner:

**Før rennet**
0 - Workflow: Opprette arrangement - før påmeldingsfrist (løpsinformasjon, klasser, startrekkefølge, antall traseer)
- [ ] Løpsinformasjon
- [ ] Bestemme aldersklasser og startrekkefølge (i GUI, følge skiforbundet sine standard aldersklasser)

1. Workflow: Generere renn (gjøres ved frist ordinær påmelding) -> skal ta maks 15 minutter
- [ ] Legge inn påmeldinger (data på standard format - typisk eksport fra min-idrett)
- [ ] Bestemme løpsklasser
- [ ] Velge hvor mange klasser som skal gå "samtidig". (2 og 2, 1 årsklasse, gir bedre plass) - 4 og 4, 2 årsklasser, gir bedre utnyttelse at tiden. Alle som går samtidig må gå samme trase.
- [ ] Innstillinger (hviletider minimum, tid til å endre trase, velge antall spor).
- [ ] Utføre seeding (importere cup-stilling) -
- [ ] Generere renn - automatisk velge mal, tilordne startnummer og plassere løpere i heat og lage kjøreplan
     - Valg: Det bør gjøres et arbeid for å lage gode maler (men det er jobbet mye med dagens maler).
     - Startnummer tilordnes samtidig som kvartfinaler genereres (etteranmeldte får egne startnummer).

2. Workflow: Etteranmelding, mange
- Etteranmelding (bulk-påmelding)

3. Utskrift på papir - direkte fra nettleser. Lage en side som også kan lagres til pdf
- [ ] Startlister (pr klasse, pr klubb)
- [ ] Kjøreplan (pr løpsklasse, totalt)

4. Manuelle justeringer
- [ ] Flytte løpere mellom heat - eks: feil ved seeding
- [ ] Justere kjøreplan (tidspunkt) - endre tidspunkt på et heat og justere alle påfølgende tilsvarende.
- [ ] Etteranmelding (enkelt på- og enkelt av-) - skal fungere helt frem til første løper i klassen starter.

**Under rennet - når første løper i klassen starter**
- [ ] Registrere start (når løper registreres inn i start-feltet)
- [ ] Registrere tidspunkt når løper passerer mål
- [ ] (sprint) Registrere rekkefølgen over mål pr heat - viktig: må optimaliseres for effektivitet
- [ ] (sprint)Publisering web+papir (heat startlister, heat resultater)
- [ ] Manuelle justeringer (resultat / heatoppsett) - løpere kom for sent til start.
- [ ] Output: Produsere resultatliste (komplett og i henhold til Skiforbundets standard) - inkludert klagefrist (+ tid for premieutdeling).

**Etter rennet**
- [ ] Publisere offisielle resultatlister pr. klasse
- [ ] Publisering web+papir (offisielle resultater) 
- [ ] Cup - beregne poengstilling

**Etter arrangementet**
- [ ] Sende faktura for påmeldinger til klubbene (lage underlag)
- [ ] Rennrapport til TD / krets (lage underlag)

**Generelle kapabiliteter**
- [ ] Roller og autentisering/tilgangsstyring (SuperAdmin/ArrangementsAdmin/Anonym) -> bruk skiforbundet sine roller.
- [ ] Integrasjon: Målfoto
- [ ] Integrasjon: Emit brikker
- [ ] Backup
- [ ] Automatisert deploy av webserver
- [ ] Avansert heat oppsett - også lage et opplegg for D-, E- og F-finaler.

Definisjoner:
- Et arrangement består av mange renn - hver løpsklasse betraktes som et renn.

Tilnærming:
- Lage automatiske funksjonelle tester - "feature test" (dette vil representere brukerhistorien) - primært mot service_metoder/API.
- Vi ønsker å automatisere mest mulig og minimere brukerinteraksjon. Standardisere oppsett.

**Comment by stigbd** [comment]
Vi bør skille mellom tre faser:
1. Før rennet,
2. Under rennet,
3. Etter rennet

**Comment by stigbd** [comment]
Forslag til metodvalg: MVP _på tvers_ av verdikjeden

Dvs lage "hello-world" implementasjon som støtter alle kritiske steg i _heile_ verdikjeden.
